src/bin/pg_autoctl/supervisor.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * src/bin/pg_autoctl/supervisor.h
+ *   Utilities to start/stop the pg_autoctl services.
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ */
+#ifndef SUPERVISOR_H
+#define SUPERVISOR_H
+
+#include <inttypes.h>
+#include <signal.h>
+
+/*
+ * Our supervisor process may retart a service sub-process when it quits,
+ * depending on the exit status and the restart policy that has been choosen:
+ *
+ * - A permanent child process is always restarted.
+ *
+ * - A temporary child process is never restarted.
+ *
+ * - A transient child process is restarted only if it terminates abnormally,
+ *   that is, with an exit code other EXIT_CODE_QUIT (zero).
+ */
+typedef enum
+{
+	RP_PERMANENT = 0,
+	RP_TEMPORARY,
+	RP_TRANSIENT
+} RestartPolicy;
+
+
+/*
+ * Supervisor restart strategy.
+ *
+ * The idea is to restart processes that have failed, so that we can stay
+ * available without external intervention. Sometimes though if the
+ * configuration is wrong or the data directory damaged beyond repair or for
+ * some reasons, the service can't be restarted.
+ *
+ * There is no magic or heuristic that can help us decide if a failure is
+ * transient or permanent, so we implement the simple thing: we restart our
+ * dead service up to 5 times in a row, and spend up to 10 seconds retrying,
+ * and stop as soon as one of those conditions is reached.
```

Comment:
This comment still explains the old behaviour.

Reply:
Removed. The per-service Restart Policy is detailed just before with the enum data type introduction so I don't think we need to explain it here again.